R/distribution_formulas.R

```diff
@@ -0,0 +1,192 @@
+#' Set up a new distribution for a data type
+#'
+#' @param data_type A string specifying the type of data that the distribution
+#'   will be fit to. Options are listed in the function call, where the first
+#'   option listed, i.e., `"index"` is the default.
+#' @param module An identifier to a C++ fleet module that is linked to the data
+#'   of interest.
+#' @param family A description of the error distribution and link function to
+#'   be used in the model.
+#' @param sd A list of length two. The first entry, `"value"`, stores the
+#'   initial values for the relevant standard deviations. The second entry,
+#'   `"estimated"` is a vector of booleans indicating whether or not
+#'   standard deviation is estimated.
+#' @export
+new_data_distribution <- function(
+  data_type = c("index", "cpue", "agecomp", "lengthcomp"),
+  module,
+  family,
+  sd = list(value = 1, estimated = FALSE)
+) {
+  data_type <- match.arg(data_type)
+  families <- c("lognormal", "gaussian", "multinomial")
+  if (family$family == "normal") {
+    stop("use family = gaussian() instead")
+  }
+  if (!(family$family %in% families)) {
+    stop("FIMS currently does not offer this distribution.")
+  }
+  if ((data_type == "agecomp" | data_type == "lengthcomp") &
+      (family$family == "lognormal" | family$family == "gaussian")) {
+    stop("Did you mean family = multinomial()?")
+  }
+  if ((data_type == "index" | data_type == "cpue") &
+      family$family == "multinomial") {
+    stop("Multinomial is not available for index or CPUE data")
+  }
+
+  if (data_type == "index" | data_type == "cpue") {
+    obs_id_name <- "observed_index_data_id"
+  }
+  if (data_type == "agecomp" | data_type == "lengthcomp") {
+    obs_id_name <- "observed_agecomp_data_id"
+  }
+  if (family$family == "lognormal") {
+    new_module <- new(TMBDlnormDistribution)
```

Comment:
When are we going to rename these to not have TMB in the name? Should I add this to the list in the M2 Quality issue #662?